streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -454,6 +479,24 @@ private void handleTasksWithoutStateUpdater(final Map<TaskId, Set<TopicPartition
                                                 final Map<TaskId, Set<TopicPartition>> standbyTasksToCreate,
                                                 final Map<Task, Set<TopicPartition>> tasksToRecycle,
                                                 final Set<Task> tasksToCloseClean) {
+        final Map<Task, Set<TopicPartition>> initialStandbyTasksToRecycle = assignInitialTasks(activeTasksToCreate, logPrefix, topologyMetadata, changelogReader);
+        final Map<Task, Set<TopicPartition>> initialStandbyTasksToUse = assignInitialTasks(activeTasksToCreate, logPrefix, topologyMetadata, changelogReader);
+        final Set<TaskId> recycledInitialTasks = new HashSet<>(initialStandbyTasksToRecycle.size() + initialStandbyTasksToUse.size());
+
+        // if this was the last local thread to receive its assignment, close all the remaining Tasks, as they are not needed
+        stateDirectory.closeInitialTasksIfLastAssginedThread();
+
+        // recycle the initial standbys to active, and remove them from the set of actives that need to be created
+        tasksToRecycle.putAll(initialStandbyTasksToRecycle);
```

Comment:
Would it be easier to just add these tasks to `TaskRegistry` `this.tasks` and everything else falls into place automatically by the existing code?

Reply:
Not sure I understand what you're suggesting here. We're essentially just adding the Tasks to the TaskRegistry, and then letting the loop below finish setting them up.

We have to do some extra book-keeping to ensure we don't create new Tasks that we already have "pending" standbys for.

Can you elaborate?

Reply:
> We have to do some extra book-keeping to ensure we don't create new Tasks that we already have "pending" standbys for.

Guess this is the part I don't see? Why do we need to expliclity add them to `tasksToRecycle` and remove from `activeTasksToCreate`? After we add them via `tasks.addStandbyTasks(initialStandbyTasksToRecycle.keySet());` the for-loop below should do everything else we need? Or what do I miss?

Reply:
Ahh yes, good point. I'm not sure how I missed that. I've simplified it in https://github.com/apache/kafka/pull/16922/commits/a1d5fba1e19a621d093b5eeb5e41db8cb588f644.

I'm not sure if we can do something similar for the `StateUpdater` version of assignment. That code is organised a little differently. I'll look into it.

Reply:
Had a look at the version for the StateUpdater, and sadly we can't just add the Standbys to the StateUpdater and let the existing code sort out the recycling etc, because the StateUpdater thread is still running in the background and attempts to start actually processing the Standbys after they've been added but before they've been recycled.